### PR TITLE
feat: allow to specify dockerfile name in `image_build_spec`

### LINKED ISSUE #2187: Allow to specify the name of the Dockerfile
### Background & motivation

By default, `ImageBuildSpec` will search for a file called `Dockerfile` in the `build_context_dir`. I would like to be able to specify the name of the Dockerfile via an optional argument.

Just like what `docker build` allows:

```bash
$ docker build --help | grep file
    -f, --file string                   Name of the Dockerfile (default: "PATH/Dockerfile")
```

Btw I'm open to work on this to get more familiar with the code base if possible :)

### Desired behaviour

Here is the new specification of `ImageBuildSpec` with an optional additional argument called `file`.

```python
image = ImageBuildSpec(
      # Name to give built image
      # MANDATORY
      image_name="kurtosistech/example-datastore-server"
      
      # Locator to build context within the Kurtosis package
      # MANDATORY
      build_context_dir="./server"

      # Name of the Dockerfile
      # By default, Kurtosis expects a Dockerfile at the root of the build context.
      # OPTIONAL
      file="mycustomfile.Dockerfile"
      
      # Stage of image build to target for multi-stage container image
      # OPTIONAL
      target_stage=""
)
```

### How important is this to you?

Nice to have; this feature would make using Kurtosis more enjoyable.

### What area of the product does this pertain to?

Other: anything not covered by the above

### PR DESCRIPTION
**Pull request by leovct**
## Description:

Allow to specify the Dockerfile name in `ImageBuildSpec`.
Resolves #2187 

- Add a new `build_file` attribute in `ImageBuildSpec`.
- Update `ImageBuildSpec` tests and add custom build file tests
- Update the `ImageBuildSpec`, `TargetStageAttr` validator to `nil` since it's an optional value.
- Update `ImageBuildSpec`docs.
- (not related) `go mod tidy`.

## Is this change user facing?
✅

### CODE REVIEW

**Comment by mieubrisse** [comment]
Thanks @leovct ! Looks like there's still a failing test here: https://app.circleci.com/pipelines/github/kurtosis-tech/kurtosis/11579/workflows/04b3d332-76f8-4d63-9cdc-b82e936f4fab/jobs/169175?invite=true#step-108-11902_114 and a merge conflict with `main`, but I'm excited where this is going!

File: core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go [hunk 1]
```diff
@@ -9,18 +12,16 @@ import (
...
 
 const (
 	ImageBuildSpecTypeName = "ImageBuildSpec"
 
 	BuiltImageNameAttr = "image_name"
 	BuildContextAttr   = "build_context_dir"
+	BuildFileAttr      = "build_file"
 	TargetStageAttr    = "target_stage"
 
-	// Currently only supports container images named Dockerfile
```
Comment (mieubrisse) [inline core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go:23]: nice; thanks for removing 💪

**Review by mieubrisse — APPROVED** [review]
Great job with this PR @leovct ; very thorough!

File: core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go [hunk 2]
```diff
@@ -45,13 +46,17 @@ func NewImageBuildSpecType() *kurtosis_type_constructor.KurtosisTypeConstructor
...
+					Name:              BuildFileAttr,
+					IsOptional:        true,
+					ZeroValueProvider: builtin_argument.ZeroValueProvider[starlark.String],
+					Validator:         nil,
+				},
 				{
 					Name:              TargetStageAttr,
 					IsOptional:        true,
 					ZeroValueProvider: builtin_argument.ZeroValueProvider[starlark.String],
-					Validator: func(value starlark.Value) *startosis_errors.InterpretationError {
```
Comment (leovct) [inline core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go]: Not sure about that @mieubrisse since I'm not very familiar with the code base

Comment (mieubrisse) [reply core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go]: oops didn't catch this; yeah let's not change the TargetStageAttr validator for now. I _think_ you should even be able to apply the same validator to the build-file arg too

Comment (leovct) [reply core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go]: Alright, let me provide some context :)

The tests for `ImageBuildSpec` were initially designed to cover scenarios where `target_stage` is specified. Given that `target_stage` is an optional argument and can also be empty, I introduced a test to address this case. However, the test did not pass. The issue lies in the fact that the validator expects `TargetStageAttr` to be non-empty, conflicting with its optional nature. To resolve this conflict, I set the validator for `TargetStageAttr` to `nil`. I followed the same approach for the newly introduced argument, `BuildFileAttr`.

Below are the logs I encountered when configuring the `TargetStageAttr` validator to `builtin_argument.NonEmptyString`:

```sh
~/kurtosis/core/server $ go test -v ./... 
....
RUN   TestKurtosisTypeConstructorSuite/TestImageBuildSpecTest
    suite_kurtosis_type_constructor_test.go:54: 
        	Error Trace:	/Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/suite_kurtosis_type_constructor_test.go:54
        	            				/Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/image_build_spec_framework_test.go:30
        	Error:      	Expected nil, but got: &starlark.EvalError{Msg:"Cannot construct 'ImageBuildSpec' from the provided arguments.\n\tCaused by: The following argument(s) could not be parsed or did not pass validation: {\"target_stage\":\"Value for 'target_stage' was an empty string. This is disallowed\"}", CallStack:starlark.CallStack{starlark.CallFrame{Name:"<toplevel>", Pos:syntax.Position{file:(*string)(0xc0000febc0), Line:1, Col:24}}, starlark.CallFrame{Name:"ImageBuildSpec", Pos:syntax.Position{file:(*string)(0x2b610a0), Line:0, Col:0}}}, cause:(*startosis_errors.InterpretationError)(0xc00029cd40)}
        	Test:       	TestKurtosisTypeConstructorSuite/TestImageBuildSpecTest
        	Messages:   	Error interpreting Starlark code. Code was: 
        	            	result = ImageBuildSpec(image_name="kurtosistech/example-datastore-server", build_context_dir="./", build_file="", target_stage="")
    mock_package_content_provider.go:519: FAIL:	GetAbsoluteLocator(string,string,string,map[string]string)
        		at: [/Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/startosis_packages/mock_package_content_provider.go:161 /Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/image_build_spec_framework_test.go:21]
    mock_package_content_provider.go:519: FAIL:	GetOnDiskAbsolutePackageFilePath(string)
        		at: [/Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/startosis_packages/mock_package_content_provider.go:325 /Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/image_build_spec_framework_test.go:26]
    mock_package_content_provider.go:519: FAIL: 0 out of 2 expectation(s) were met.
        	The code you are testing needs to make 2 more call(s).
        	at: [/Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/startosis_packages/mock_package_content_provider.go:519 /Users/leovct/.asdf/installs/golang/1.20.14/go/src/testing/testing.go:1150 /Users/leovct/.asdf/installs/golang/1.20.14/go/src/testing/testing.go:1328 /Users/leovct/.asdf/installs/golang/1.20.14/go/src/testing/testing.go:1570 /Users/leovct/.asdf/installs/golang/1.20.14/go/src/runtime/panic.go:522 /Users/leovct/.asdf/installs/golang/1.20.14/go/src/testing/testing.go:980 /Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/suite_kurtosis_type_constructor_test.go:54 /Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/image_build_spec_framework_test.go:30]
```

Comment (mieubrisse) [reply core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go]: Ah, thanks for the context @leovct ! Let me tag in @tedim52 who has context on target stage, to figure out which is the correct one so we can get your PR merged ASAP :)

Comment (tedim52) [reply core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go]: Hey @leovct ! Thanks for adding a test for target stage. 

The behavior of the validator function is that if the argument is set (optional or not), the validator function is triggered. So if `target_stage` is not set at all, thats fine because it's optional, but if it is set, then we require it not to be the empty string. This behavior should be the exact same for `build_file`. 

To get that behavior, you can add back the original validator for `target_stage` and add it to `build_file` as well. Then for the tests, add a test where `target_stage` is empty and an error occurs (tests the validation) and a test where `target_stage` is non empty (should succeed) and repeat for `build_file`. LMK if this makes sense. Thanks again for improving our testsuite : )

Comment (leovct) [reply core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go]: That makes sense, thank you! I [reverted](https://github.com/kurtosis-tech/kurtosis/pull/2223/commits/8fab5c37316aded553c3f63278baa6fc8219cafa) the change on the validator functions and fixed the tests :)

**Comment by mieubrisse** [comment]
Cloudflare is having issues right now, so I'm going to admin-merge this (the rest of the checks passed)

**Comment by mieubrisse** [comment]
Tracking: seems like there's flakiness in our CI pipeline (tests passed in the PR, but failed in the merge queue): https://app.circleci.com/pipelines/github/kurtosis-tech/kurtosis/11662/workflows/326000e1-44d9-43d6-b366-d14966c4b808/jobs/170803

@laurentluce to investigate
